Add stdout and infinite targets

jeffreylovitz opened the pull request:
My use case for grammarinator involves setting up a CI environment that continuously receives fuzzed queries. To this end, this PR introduces support for the flags `-o stdout` and `--infinite`.

These flags are not expected to work in parallelized environments, but I think this is a useful alternative to tests written to disk. Let me know if this idea is interesting to you and, if so, what this PR needs!

docfate111 commented:
Please merge this. This is exactly what is needed.

renatahodovan commented on grammarinator/generate.py:
```diff
@@ -289,8 +298,12 @@ def restricted_float(value):
             with Pool(args.jobs) as pool:
```
Infinite test generation should be supported in multiprocess mode as well.

renatahodovan commented on grammarinator/generate.py:
```diff
@@ -130,8 +134,11 @@ def create_new_test(self, index):
+            with codecs.open(test_fn, 'w', self.encoding) as f:
+                f.write(self.serializer(tree.root))
+        else:
+            print(self.serializer(tree.root))
```
If grammarinator runs in multiprocess model, how will be ensured that the result of various threads won't be interleaved on the stdout? Maybe a lock should be shared between the threads and enable to print only for the holder of the lock.

jeffreylovitz commented on grammarinator/generate.py:
```diff
@@ -286,10 +307,11 @@ def restricted_float(value):
-                pool.map(generator, range(args.n))
+            shared_lock = Lock()
+            with Pool(args.jobs, initializer=generator.set_lock, initargs=(shared_lock,)) as pool:
+                pool.map(generator, count(0) if args.n == 'inf' else range(args.n))
```
This line causes aggressive memory consumption leading to a crash if `inf` is set. @renatahodovan Do you have any idea what could be wrong?

jeffreylovitz commented:
@renatahodovan Thanks for the review!

I've addressed your comments, but there is one outstanding issue I indicated here - https://github.com/renatahodovan/grammarinator/pull/32#discussion_r608173705 .

renatahodovan reviewed (changes requested):
@jeffreylovitz  Thanks for the update and I'm sorry for the late reply. Although the patch needs some further refinements but I think that the solution is close.

jeffreylovitz commented on grammarinator/generate.py:
```diff
@@ -286,11 +300,16 @@ def restricted_float(value):
         else:
-            for i in range(args.n):
-                generator(i)
+            lock = Manager().Lock() # Build an uncontested lock
```
@renatahodovan We build a lock for the single-threaded case so that the [with lock](https://github.com/renatahodovan/grammarinator/pull/32/files#diff-2f19df232ab8dd27c19e9e80fe050965b0a7a7a43ed0503d06ce176486fa715dR145) call doesn't trip on `None`.

I think this is preferable to having null handling at that point, but let me know if you'd prefer a different approach!

renatahodovan commented on grammarinator/generate.py:
```diff
@@ -12,9 +12,12 @@
+from itertools import count
+from functools import partial
+from contextlib import contextmanager
+from multiprocessing import Pool, Manager
```
Please sort imports alphabetically:

```py
from argparse import ArgumentParser, ArgumentTypeError, SUPPRESS
from contextlib import contextmanager
from functools import partial
from itertools import count
from math import inf
from multiprocessing import Manager, Pool
from os.path import abspath, basename, dirname, isdir, join, splitext
from shutil import rmtree
```

renatahodovan commented on grammarinator/generate.py:
```diff
@@ -121,7 +130,7 @@ def create_new_test(self, index):
             logger.warning('Test generation failed.', exc_info=e)
             return self.create_new_test(index)
```
The `lock` parameter should be added here, as well.

```py
return self.create_new_test(index, lock)
```

renatahodovan commented on grammarinator/generate.py:
```diff
@@ -286,10 +301,14 @@ def restricted_float(value):
-                pool.map(generator, range(args.n))
+            with Manager() if not args.out else nullcontext() as manager:
+                if not args.out:
+                    generator = partial(generator, lock=manager.Lock())
```
pylint [complains](https://github.com/renatahodovan/grammarinator/runs/2838943063?check_suite_focus=true) about this line, but it seems that it's a [known pylint issue](https://github.com/PyCQA/pylint/issues/3313), so we can just ignore it. This should work, but you can validate it by running tox locally:

```py
generator = partial(generator, lock=manager.Lock())  # pylint: disable=no-member
```

renatahodovan reviewed (changes requested):
After fixing the aforementioned small issues, please squash your commits and rebase them to the latest master.

renatahodovan reviewed (approved):
Lgtm. Thanks!

jeffreylovitz commented:
Great! Thanks for authoring this project, it is highly useful.
